File: python/snake_game/input_handler.py
"""maps pygame key events to game commands"""
import pygame

from .snake import DOWN, LEFT, RIGHT, UP

DIRECTION_KEYS = {
    pygame.K_LEFT: LEFT, pygame.K_a: LEFT,
    pygame.K_RIGHT: RIGHT, pygame.K_d: RIGHT,
    pygame.K_UP: UP, pygame.K_w: UP,
    pygame.K_DOWN: DOWN, pygame.K_s: DOWN,
}


class InputHandler:
    def poll(self):
        commands = []
        for event in pygame.event.get():
            if event.type == pygame.QUIT:
                commands.append(('quit', None))
            elif event.type == pygame.KEYDOWN:
                if event.key in DIRECTION_KEYS:
                    commands.append(('direction', DIRECTION_KEYS[event.key]))
                elif event.key == pygame.K_TAB:
                    commands.append(('toggle_ai', None))
                elif event.key == pygame.K_h:
                    commands.append(('toggle_ai_mode', None))
        return commands

File: cpp/src/input_handler.cpp
#include "input_handler.hpp"

#include <raylib.h>

std::vector<Command> InputHandler::Poll() const {
    std::vector<Command> commands;

    if (WindowShouldClose()) commands.push_back(Command{CommandType::Quit});

    if (IsKeyPressed(KEY_LEFT) || IsKeyPressed(KEY_A)) {
        commands.push_back(Command{CommandType::Direction, Direction::Left});
    }
    if (IsKeyPressed(KEY_RIGHT) || IsKeyPressed(KEY_D)) {
        commands.push_back(Command{CommandType::Direction, Direction::Right});
    }
    if (IsKeyPressed(KEY_UP) || IsKeyPressed(KEY_W)) {
        commands.push_back(Command{CommandType::Direction, Direction::Up});
    }
    if (IsKeyPressed(KEY_DOWN) || IsKeyPressed(KEY_S)) {
        commands.push_back(Command{CommandType::Direction, Direction::Down});
    }
    if (IsKeyPressed(KEY_TAB)) commands.push_back(Command{CommandType::ToggleAi});
    if (IsKeyPressed(KEY_H)) commands.push_back(Command{CommandType::ToggleAiMode});

    return commands;
}
